Stock Market Comparison Analysis is a methodical examination of multiple stocks or financial assets within the stock market. It involves analyzing the performance of various stocks or assets to gain insights into how they have fared relative to each other and the broader market. It helps investors, financial analysts, and decision-makers make informed investment decisions.

Below is the process:
- Gather historical stock price data for the companies or financial instruments you want to compare.
- Clean and preprocess the collected data.
- Create relevant features or variables that can help in the comparison.
- Ensure that the data for different stocks is aligned in terms of time.
- Use data visualization techniques to present the comparison effectively.
- Apply statistical methods to identify patterns or relationships between the stocks being compared.
- Compare the performance of the selected stocks or financial instruments against relevant benchmarks, such as market indices like the S&P 500.

# Import the necessary Python libraries

In [1]:
!pip install yfinance

In [2]:
import pandas as pd
import yfinance as yf
import plotly.io as pio
import plotly.graph_objects as go
pio.templates.default = "plotly_white"

# Collect stock price data:

In [3]:
# Define the tickers for Apple and Google
apple_ticker = 'AAPL'
google_ticker = 'GOOGL'

# Define the date range for the last quarter
start_date = '2023-07-01'
end_date = '2023-09-30'

# Fetch historical stock price data using yfinance
apple_data = yf.download(apple_ticker, start=start_date, end=end_date)
google_data = yf.download(google_ticker, start=start_date, end=end_date)

[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed


In [4]:
apple_data.head()

Open        High         Low       Close   Adj Close  \
Date                                                                     
2023-07-03  193.779999  193.880005  191.759995  192.460007  191.221954   
2023-07-05  191.570007  192.979996  190.619995  191.330002  190.099228   
2023-07-06  189.839996  192.020004  189.199997  191.809998  190.576126   
2023-07-07  191.410004  192.669998  190.240005  190.679993  189.453400   
2023-07-10  189.259995  189.990005  187.039993  188.610001  187.396698   

              Volume  
Date                  
2023-07-03  31458200  
2023-07-05  46920300  
2023-07-06  45094300  
2023-07-07  46778000  
2023-07-10  59922200

This data will serve as the basis for various financial analyses and comparisons, such as calculating returns, volatility, and other metrics to assess the performance and risk associated with these stocks.

# Daily returns of Apple and Google

## Calculate daily returns

In [5]:
# Calculate daily returns
apple_data['Daily_Return'] = apple_data['Adj Close'].pct_change()
google_data['Daily_Return'] = google_data['Adj Close'].pct_change()

In [6]:
apple_data.head()

Open        High         Low       Close   Adj Close  \
Date                                                                     
2023-07-03  193.779999  193.880005  191.759995  192.460007  191.221954   
2023-07-05  191.570007  192.979996  190.619995  191.330002  190.099228   
2023-07-06  189.839996  192.020004  189.199997  191.809998  190.576126   
2023-07-07  191.410004  192.669998  190.240005  190.679993  189.453400   
2023-07-10  189.259995  189.990005  187.039993  188.610001  187.396698   

              Volume  Daily_Return  
Date                                
2023-07-03  31458200           NaN  
2023-07-05  46920300     -0.005871  
2023-07-06  45094300      0.002509  
2023-07-07  46778000     -0.005891  
2023-07-10  59922200     -0.010856

In [7]:
google_data.head()

Open        High         Low       Close   Adj Close  \
Date                                                                     
2023-07-03  119.239998  120.190002  118.820000  119.900002  119.762550   
2023-07-05  119.239998  122.610001  119.230003  121.750000  121.610428   
2023-07-06  119.800003  120.300003  118.400002  120.110001  119.972313   
2023-07-07  120.099998  121.050003  119.400002  119.480003  119.343033   
2023-07-10  118.300003  118.309998  116.139999  116.449997  116.316498   

              Volume  Daily_Return  
Date                                
2023-07-03  14467900           NaN  
2023-07-05  27584800      0.015430  
2023-07-06  24745200     -0.013470  
2023-07-07  21692600     -0.005245  
2023-07-10  35315200     -0.025360

## visualize the daily returns:

In [8]:
# Create a figure to visualize the daily returns
fig = go.Figure()

fig.add_trace(go.Scatter(x=apple_data.index, y=apple_data['Daily_Return'],
                         mode='lines', name='Apple', line=dict(color='blue')))
fig.add_trace(go.Scatter(x=google_data.index, y=google_data['Daily_Return'],
                         mode='lines', name='Google', line=dict(color='green')))

fig.update_layout(title='Daily Returns for Apple and Google (Last Quarter)',
                  xaxis_title='Date', yaxis_title='Daily Return',
                  legend=dict(x=0.02, y=0.95))

fig.show()

## Analyze the cumulative returns of Apple and Google for the last quarter

In [9]:
# Calculate cumulative returns for the last quarter
apple_cumulative_return = (1 + apple_data['Daily_Return']).cumprod() - 1
google_cumulative_return = (1 + google_data['Daily_Return']).cumprod() - 1

# Create a figure to visualize the cumulative returns
fig = go.Figure()

fig.add_trace(go.Scatter(x=apple_cumulative_return.index, y=apple_cumulative_return,
                         mode='lines', name='Apple', line=dict(color='blue')))
fig.add_trace(go.Scatter(x=google_cumulative_return.index, y=google_cumulative_return,
                         mode='lines', name='Google', line=dict(color='green')))

fig.update_layout(title='Cumulative Returns for Apple and Google (Last Quarter)',
                  xaxis_title='Date', yaxis_title='Cumulative Return',
                  legend=dict(x=0.02, y=0.95))

fig.show()

In [10]:
apple_cumulative_return

Date
2023-07-03         NaN
2023-07-05   -0.005871
2023-07-06   -0.003377
2023-07-07   -0.009249
2023-07-10   -0.020004
                ...   
2023-09-25   -0.083873
2023-09-26   -0.105309
2023-09-27   -0.113270
2023-09-28   -0.111917
2023-09-29   -0.109211
Name: Daily_Return, Length: 63, dtype: float64

Here, we first calculated the cumulative returns for both Apple and Google over the last quarter. Cumulative returns represent the total percentage change in the stock’s value over a given period, considering the compounding effect of daily returns. Then, we compared the investment performance of Apple and Google during the specified period, showing which stock had higher or lower cumulative returns over that time frame. We can see that Google has higher cumulative returns compared to Apple over the last quarter.

## Calculate historical volatility (standard deviation of daily returns)

In [11]:
apple_volatility = apple_data['Daily_Return'].std()
google_volatility = google_data['Daily_Return'].std()

In [12]:
apple_volatility

0.013463543901903382

In [13]:
# Create a figure to compare volatility
fig1 = go.Figure()
fig1.add_bar(x=['Apple', 'Google'], y=[apple_volatility, google_volatility],
             text=[f'{apple_volatility:.4f}', f'{google_volatility:.4f}'],
             textposition='auto', marker=dict(color=['blue', 'green']))

fig1.update_layout(title='Volatility Comparison (Last Quarter)',
                   xaxis_title='Stock', yaxis_title='Volatility (Standard Deviation)',
                   bargap=0.5)
fig1.show()

We first calculated the historical volatility for both Apple and Google stocks. Volatility is a measure of how much the stock’s price fluctuates over time. In this case, we are calculating the standard deviation of daily returns to measure the volatility. Then we visualized the calculated volatility to assess and compare the volatility or risk associated with both Apple and Google stocks during the specified period. We can see that Google’s volatility is higher than Apple’s.

It indicates that Google’s stock price experienced larger price fluctuations or greater price variability over the last quarter. Here’s what this difference in volatility may indicate:
- Google’s stock is considered riskier compared to Apple. Investors generally associate higher volatility with higher risk because it implies that the stock price can change significantly in a short period.
- Google’s stock may be more sensitive to market conditions, economic factors, or company-specific news and events. This heightened sensitivity can result in larger price swings.
- Traders and investors with a higher risk tolerance might find Google’s stock appealing if they are looking for opportunities to profit from short-term price movements.

# Compare the stock market of Google and Apple according to the stock market benchmark:

In [14]:
market_data = yf.download('^GSPC', start=start_date, end=end_date)  # S&P 500 index as the market benchmar
market_data

[*********************100%***********************]  1 of 1 completed


Open         High          Low        Close    Adj Close  \
Date                                                                          
2023-07-03  4450.479980  4456.459961  4442.290039  4455.589844  4455.589844   
2023-07-05  4442.040039  4454.060059  4436.609863  4446.819824  4446.819824   
2023-07-06  4422.620117  4422.620117  4385.049805  4411.589844  4411.589844   
2023-07-07  4404.540039  4440.390137  4397.399902  4398.950195  4398.950195   
2023-07-10  4394.229980  4412.600098  4389.919922  4409.529785  4409.529785   
...                 ...          ...          ...          ...          ...   
2023-09-25  4310.620117  4338.509766  4302.700195  4337.439941  4337.439941   
2023-09-26  4312.879883  4313.009766  4265.979980  4273.529785  4273.529785   
2023-09-27  4282.629883  4292.069824  4238.629883  4274.509766  4274.509766   
2023-09-28  4269.649902  4317.270020  4264.379883  4299.700195  4299.700195   
2023-09-29  4328.180176  4333.149902  4274.859863  4288.049805  4288.049805   

                Volume  
Date                    
2023-07-03  2034280000  
2023-07-05  3482620000  
2023-07-06  3682020000  
2023-07-07  3630480000  
2023-07-10  3429600000  
...                ...  
2023-09-25  3195650000  
2023-09-26  3472340000  
2023-09-27  3875880000  
2023-09-28  3846230000  
2023-09-29  3865960000  

[63 rows x 6 columns]

In [15]:
# Calculate daily returns for both stocks and the market
apple_data['Daily_Return'] = apple_data['Adj Close'].pct_change()
google_data['Daily_Return'] = google_data['Adj Close'].pct_change()
market_data['Daily_Return'] = market_data['Adj Close'].pct_change()

In [16]:
# Calculate Beta for Apple and Google
cov_apple = apple_data['Daily_Return'].cov(market_data['Daily_Return'])
var_market = market_data['Daily_Return'].var()

beta_apple = cov_apple / var_market

cov_google = google_data['Daily_Return'].cov(market_data['Daily_Return'])
beta_google = cov_google / var_market

# Compare Beta values
if beta_apple > beta_google:
    conclusion = "Apple is more volatile (higher Beta) compared to Google."
else:
    conclusion = "Google is more volatile (higher Beta) compared to Apple."

# Print the conclusion
print("Beta for Apple:", beta_apple)
print("Beta for Google:", beta_google)
print(conclusion)

Beta for Apple: 1.2257471653319827
Beta for Google: 1.5303072232101351
Google is more volatile (higher Beta) compared to Apple.


In the above code, we are assessing how sensitive Apple and Google stocks are to overall market movements, providing insights into their relative volatility and risk about the broader U.S. stock market represented by the S&P 500 index.

The Standard & Poor’s 500, often referred to as the S&P 500, is a widely recognized stock market index in the United States. The S&P 500 index includes 500 of the largest publicly traded companies in the United States, chosen for their market capitalization, liquidity, and industry representation. These companies span various sectors of the U.S. economy and provide a comprehensive view of the health and performance of the stock market.

In the above output, the beta value for Apple is approximately 1.2257. This beta value suggests that Apple’s stock is estimated to be approximately 22.57% more volatile or sensitive to market movements (as represented by the S&P 500 index) compared to the overall market. The beta value for Google is approximately 1.5303. This beta value suggests that Google’s stock is estimated to be approximately 53.03% more volatile or sensitive to market movements.

A beta greater than 1 suggests that a stock tends to be more volatile than the market. In this case, both Apple and Google have beta values greater than 1, indicating that they are expected to be more volatile and sensitive to market movements. Google’s higher beta value (1.5303) compared to Apple’s (1.2257) suggests that Google’s stock is estimated to have a higher degree of market sensitivity or risk compared to Apple. Investors should consider this information when making investment decisions, as higher-beta stocks can provide greater potential for returns but also carry a higher level of risk.